website/oauth/models/__init__.py

```diff
@@ -0,0 +1,261 @@
+import abc
+import datetime
+import logging
+
+from flask import request
+from modularodm import fields
+from modularodm import Q
+from modularodm.exceptions import NoResultsFound
+from requests_oauthlib import OAuth1Session
+from requests_oauthlib import OAuth2Session
+
+from framework.exceptions import PermissionsError
+from framework.mongo import ObjectId
+from framework.mongo import StoredObject
+from framework.sessions import get_session
+from website import settings
+from website.oauth.utils import PROVIDER_LOOKUP
+from website.util import web_url_for
+
+
+logger = logging.getLogger(__name__)
+
+OAUTH1 = 1
+OAUTH2 = 2
+
+
+class ExternalAccount(StoredObject):
+    _id = fields.StringField(default=lambda: str(ObjectId()))
+
+    oauth_key = fields.StringField()
+    oauth_secret = fields.StringField()
+    refresh_token = fields.StringField()
+    expires_at = fields.DateTimeField()
+
+    scopes = fields.StringField(list=True, default=lambda: list())
+
+    # The `name` of the service
+    provider = fields.StringField(required=True)
+
+    # The unique, persistent ID on the remote service.
+    provider_id = fields.StringField()
+
+    display_name = fields.StringField()
+
+    def __repr__(self):
+        return '<ExternalAccount: {}/{}>'.format(self.provider,
+                                                 self.provider_id)
+
+
+class ExternalProviderMeta(abc.ABCMeta):
+
+    def __init__(cls, name, bases, dct):
+        super(ExternalProviderMeta, cls).__init__(name, bases, dct)
+        if not isinstance(cls.short_name, abc.abstractproperty):
+            PROVIDER_LOOKUP[cls.short_name] = cls
+
+
+class ExternalProvider(object):
+    """
+
+    """
+
+    __metaclass__ = ExternalProviderMeta
+
+    account = None
+    _oauth_version = OAUTH2
+
+    def __repr__(self):
+        return '<{name}: {status}>'.format(
+            name=self.__class__.__name__,
+            status=self.account.provider_id if self.account else 'anonymous'
+        )
+
+    @abc.abstractproperty
+    def auth_url_base(self):
+        """The base URL to begin the OAuth dance"""
+        raise NotImplementedError
+
+    @property
```

Comment:
This method does a lot of work--fetches current session, sends HTTP requests--so it may be misleading to make this a property. Maybe we should make this a method (`fetch_auth_url()`)?